57 molecular dynamics support for fhiaims

JFRudzinski opened the pull request:
Changes based on this one test example: https://github.com/nomad-coe/electronic-parsers/blob/develop/tests/data/fhiaims/HO_md/H2O_periodic_MD.out. We need to work with users to generalize to other scenarios.

JFRudzinski committed:
added basic md support for fhiaims

JFRudzinski committed:
int casting for step quantity

JFRudzinski committed:
added system and calculation tests for md fhiaims

JFRudzinski committed:
added workflow tests for md fhiaims

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -40,7 +40,13 @@
 from nomad.datamodel.metainfo.simulation.workflow import (
     SinglePoint as SinglePoint2, GeometryOptimization as GeometryOptimization2,
-    MolecularDynamics as MolecularDynamics2)
+    MolecularDynamics as MolecularDynamics2, MolecularDynamicsMethod, MolecularDynamicsResults,
```
Not a fan of keeping naming things `...2`. Do you mind changing it for all these workflows and in the corresponding `parse_workflow`?

JFRudzinski replied:
The labeling is necessary when there are duplicate imports. In this case though it all has to do with the old vs. new workflow schema. I had to cover both so that the GUI features would work properly now. A couple of the "1" labels are actually not duplicates, but I wanted to make it easy to remove once the old workflow section is deprecated.

JosePizarro3 replied:
In that case, I would define the 2 without numbering, and the 1 as old, e.g. `ThermostatParameters as ThermostatParametersOld`.

I am really not a friend of these other naming by numberings 😄

JFRudzinski replied:
ok, I agree :+1: . I changed the ones that I added myself. But there are a couple other "2"'s that were added by Alvin. I do not want to change these unless he agrees. @ladinesa: Is it ok if I change "SinglePoint2" to "SinglePoint" and "GeometryOptimization2" to "GeometryOptimization" here? (There are no instances of the original names anywhere in this parser).

ladinesa replied:
we could simply import workflow and workflow2

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -40,7 +40,13 @@
+
+# for old MD workflow
+from nomad.datamodel.metainfo.workflow import (
+    ThermostatParameters as ThermostatParameters1, IntegrationParameters as IntegrationParameters1,
```
Same here. Why the naming with 1 and 2 at the end?

JFRudzinski replied:
see above.

JosePizarro3 replied:
I saw it is resolved by changing to `...Old`. 

Thanks.

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -448,6 +488,55 @@ def str_to_gw_scf(val_in):
+        molecular_dynamics_quantities = [
+            Quantity(
+                'molecular_dynamics_ensemble',
+                r' *Running\s*Born-Oppenheimer\s*molecular\s*dynamics\s*in\s*([A-Z]{3})\s*ensemble',
```
can't be ensemble and thermostat in the same regex?

JFRudzinski replied:
Yes, in principal. But then I would have to write a really specific str_operation function to extract the 2 quantities with appropriate keys, so I thought it was just more straightforward this way. Do you think that is better? Or I could also not fully understand what you have in mind?

JosePizarro3 replied:
Well it is just a list. Then later you assign to [0] and [1].

JFRudzinski replied:
ok, changed this to a single regex.

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -448,6 +488,55 @@ def str_to_gw_scf(val_in):
 
+        molecular_dynamics_quantities = [
+            Quantity(
+                'molecular_dynamics_ensemble',
```
what about changing the naming to "md_xxx"? Or do you prefer to write fully molecular_dynamics?

JFRudzinski replied:
I don't really have a huge preference. I lean towards trying to be more verbose in general, especially cause this isn't a md-based software, so maybe the abbreviation will not be clear to some people.

JosePizarro3 replied:
Being inside `molecular_dynamics_quantities` already leaves it clear, in my opinion, but like you want.

JFRudzinski replied:
ok, shortened these to `md_xxx`

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -448,6 +488,55 @@ def str_to_gw_scf(val_in):
+                rf'{re_n} *Advancing structure using Born-Oppenheimer Molecular Dynamics:\s*{re_n}'
+                rf' *Complete information for previous time-step:'
+                rf'([\s\S]+?)((?:{re_n}{re_n}|\| Nose-Hoover Hamiltonian\s*:\s*[Ee\d\.\-\+]+ eV))',
+                str_operation=str_to_md_calculation_info, repeats=False, convert=False),
```
you could define `sub_parser=TextParser(quantities=[...])` instead of defining the str_operation (which sometimes is pretty annoying).

The quantities you must define are just the ones appearing after the | and you can give them names, e.g. 'time_step_number'.

This is the point were I told you you could decide to store as much as possible in a code-specific variable of type=JSON by simply passing:
`sec_scc.x_fhi_aims_molecular_dynamics_calculation_info = self.out_parser.get('control_inout').get('molecular_dynamics_calculation_info').results`
and of course defining this `x_fhi_aims_molecular_dynamics_calculation_info` in the metainfo. Then it is a mater of mapping this into the corresponding metainfo you want to keep.

JosePizarro3 replied:
By parsing the json you can forget about the mapping:
            'Thermostat value ln(s)': 'x_fhi_aims_thermostat_value_lns',
            'Thermostat momentum pi': 'x_fhi_aims_thermostat_momentum_pi',

Similarly, you could do the same for th energy_map quantities.

JFRudzinski replied:
I would like to implement this, but I think I need a concrete example to follow. Otherwise, perhaps you can walk me through it next week?

JosePizarro3 replied:
Ok, let's keep it for a casual chat. 

I was telling this, because I have the feeling that mappings can be done in one shot, instead of having to define in the `__init__` all the possible mappings.

JFRudzinski replied:
Implemented JSON storage for most of the code-dependent quantities.

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -763,6 +851,23 @@ def __init__(self):
+            'Total energy (el.+nuc.)': 'x_fhi_aims_total_energy_el_and_nuc',
+            'Nose-Hoover Hamiltonian': 'x_fhi_aims_nosehoover_hamiltonian'
+        }
+        self._md_methods_map = {
```
No need of this mapping. See below.

JFRudzinski replied:
please see my comment below.

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -1290,8 +1453,40 @@ def parse_workflow(self):
+            workflow.method.integration_timestep = control_inout.get('molecular_dynamics_timestep')
+
+            sec_thermostat_parameters = workflow.method.m_create(ThermostatParameters)
+            sec_thermostat_parameters.thermostat_type = self._md_methods_map.get(control_inout.get('molecular_dynamics_thermostat'))
```
`sec_thermostat_parameters.thermostat_type = control_inout.get('molecular_dynamics_thermostat').lower().replace('-', '_')`

JFRudzinski replied:
I am using the mapping because I have enforced enum values for many of the MD methods, and only when there is some reference paper that can be linked. This is essential for normalization because sometimes the same underlying method is used in different codes with different names. So, I don't want to just put any text that is found there.

We could however make a code-specific quantity for this in this way. But I guess let's first implement the JSON suggestion you gave above.

JosePizarro3 replied:
I see. You need more examples. Can you ask this user whether they have implemented other thermostat types? If not, you don't need the mapping; if so, then you need to extend the mapping.

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -1290,8 +1453,40 @@ def parse_workflow(self):
+            workflow = MolecularDynamics2(
+                method=MolecularDynamicsMethod(), results=MolecularDynamicsResults())
+
+            control_inout = self.out_parser.get('control_inout')
```
Error handling: `if control_inout`

JFRudzinski replied:
fixed, thanks

JosePizarro3 reviewed (commented):
Pretty good. Just a few minor changes and some comments in case you want to play further with the regex.

But I'd just change the naming of the imports, and try not to keep workflow1, workflow2, ..., workflowN. Or is there any reason I am overlooking on why this is important?

JosePizarro3 commented:
Ah, lastly: did you check that test_fhiaimsparser passes, right?

JFRudzinski commented:
> 

yes, the fhiaims tests pass

JFRudzinski commented:
> ming of the imports, and try not to keep workflow1, workflow2, ..., workflowN. Or is there any reason I am overlooking on why this is important?

Thanks for your comments!

JFRudzinski committed:
error handling for control_inout

JosePizarro3 commented:
@JFRudzinski keep in mind that #60 has been merged. Please, don't forget to rebase, as there were some minor changes in FHIaims.

Let me know if you need help with the rebasing :smile:

ladinesa commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -372,6 +378,40 @@ def str_to_gw_scf(val_in):
+            try:
+                val = ureg(val_in.replace('^', '**'))
+            except Exception:
+                self.logger.warning(rf'Problem parsing some units from .out file, could not convert: {val_in}')
```
we should avoid logging specfic info as this will overload the logger. you can pass details instead, details={'value': val_in}

JFRudzinski replied:
done.

ladinesa commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -1256,6 +1364,61 @@ def parse_section(section):
 
+            # step, time, temperature info + additional energies from molecular dynamics
+            md_info = section.get('molecular_dynamics_calculation_info', None)
+            if md_info is not None:
```
you can set default to {} so you do not need the extra if md_info is not None

JFRudzinski replied:
done.

JosePizarro3 replied:
I think it is still there if github is not trolling me.

You can delete this line 1378 after the default {}.

JFRudzinski replied:
Oops. Thank you. It is done for real now.

ladinesa commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -1256,6 +1364,61 @@ def parse_section(section):
+            md_info = section.get('molecular_dynamics_calculation_info', None)
+            if md_info is not None:
+                for key, val in md_info.items():
+                    if key in self._md_calculation_map:
```
you can do metainfo_key = self._md_calculation_map.get(key, self._md_calculation_energy_map.get(key)) so you do not need the additional if

JFRudzinski replied:
Not sure if I am understanding your suggestion. The set attribute code is different for the cases that key is in calculation_map or energy map...so I need to distinguish these 2 cases.

ladinesa replied:
you can put md_calculation_map and md_calculation_energy map together. is there a reason so split them?

JFRudzinski replied:
yes, they are being stored in different sections, directly under calculation, and under calculation->energy, respectively. Also, the items under energy are being stored as energy entries.

ladinesa commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -1256,6 +1364,61 @@ def parse_section(section):
+                            try:
+                                setattr(sec_energy, metainfo_key, EnergyEntry(value=val))
+                            except Exception:
+                                self.logger.warn(
```
can you please update this I think warn is deprecated and also data.

JFRudzinski replied:
Updated to warning throughout. Is this correct:               
`self.logger.warning( 'Error setting md calculation energy metainfo.', details={key: metainfo_key})` with details instead of data?

ladinesa replied:
I think so but please test it.

JFRudzinski replied:
tested.

ladinesa commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -1256,6 +1364,61 @@ def parse_section(section):
+                try:
+                    setattr(sec_energy, metainfo_key, EnergyEntry(value=kinetic_energy))
+                except Exception:
+                    self.logger.warn(
```
also here

JFRudzinski replied:
done.

ladinesa commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -1256,6 +1364,61 @@ def parse_section(section):
+                    mass = label_to_mass[atom_labels[i_atom]]
+                    for dim in range(3):
+                        kinetic_energy += 0.5 * mass * (velocities[i_atom][dim]) ** 2
+                metainfo_key = 'kinetic'
```
no need to defined metainfo_key here?

JFRudzinski replied:
done

ladinesa commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -1256,6 +1364,61 @@ def parse_section(section):
+            label_to_mass = {atom_params.label: atom_params.mass for atom_params in sec_scc.method_ref.atom_parameters}
+            velocities = sec_atoms.velocities
+            atom_labels = sec_atoms.get('labels')
+            if velocities is not None:
```
not sure but maybe put also atom_labels is not None. Actually veloctities and atom_labels should match so maybe put [] as default for velocities and atom_labels then do if len(velocties) > 0 and len(velocities) == len(atom_labels)

JFRudzinski replied:
no longer relevant. Removed this part.

ladinesa commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -1256,6 +1364,61 @@ def parse_section(section):
+                    try:
+                        setattr(sec_energy, metainfo_key, EnergyEntry(value=potential_energy))
+                    except Exception:
+                        self.logger.warn(
```
also here

JFRudzinski replied:
done.

JFRudzinski committed:
renamed imports

JFRudzinski commented:
> @JFRudzinski keep in mind that #60 has been merged. Please, don't forget to rebase, as there were some minor changes in FHIaims.
> 
> Let me know if you need help with the rebasing smile

rebasing was no problem :+1:

JFRudzinski committed:
adding generic x_fhiaims_MD_controlin

JFRudzinski committed:
remove MD metainfo again for rebasing

JFRudzinski committed:
addressed review comments and added more thermostat options from manual

JFRudzinski commented:
@JosePizarro3 @ladinesa : I made quite a few adjustments to the code since you reviewed it, so you might want to have another quick look. But I should have addressed all your comments I think.

JosePizarro3 commented:
Ok, done with the review. Thanks a lot, it is looking much better (we are slowly making the FHIaims parser good :slightly_smiling_face: ).

Btw, I see there is something strange with the rebasing: why github is marking changes in all the other parsers? ~~I thought once you rebase it will be a simple moving of the tip, hence not showing the changes w.r.t develop.~~ My bad, I see that you rebased into this branch. Is it not safer to rebase this in develop and move the tip? @ladinesa ?

ladinesa commented:
> Ok, done with the review. Thanks a lot, it is looking much better (we are slowly making the FHIaims parser good slightly_smiling_face ).
> 
> Btw, I see there is something strange with the rebasing: why github is marking changes in all the other parsers? ~I thought once you rebase it will be a simple moving of the tip, hence not showing the changes w.r.t develop.~ My bad, I see that you rebased into this branch. Is it not safer to rebase this in develop and move the tip? @ladinesa ?

you should always rebase to the target branch.

JFRudzinski committed:
more review changes

JFRudzinski commented:
> Ok, done with the review. Thanks a lot, it is looking much better (we are slowly making the FHIaims parser good slightly_smiling_face ).
> 
> Btw, I see there is something strange with the rebasing: why github is marking changes in all the other parsers? ~I thought once you rebase it will be a simple moving of the tip, hence not showing the changes w.r.t develop.~ My bad, I see that you rebased into this branch. Is it not safer to rebase this in develop and move the tip? @ladinesa ?

Ok, rebased properly now.

JFRudzinski commented:
@ladinesa, @JosePizarro3 - not sure where we are at on this. Can I merge it? 

@ladinesa - what about the small corresponding changes to nomad-FAIR, did you look at that MR already?

JosePizarro3 commented:
> @ladinesa, @JosePizarro3 - not sure where we are at on this. Can I merge it?
> 
> @ladinesa - what about the small corresponding changes to nomad-FAIR, did you look at that MR already?

From my side everything is fine. Please, rebase again to avoid issues (I touched the FHIaims parser again and better be safe than sorry :slightly_smiling_face: ).

ladinesa commented:
Freel free to merge.
